xla/service/gpu/dynamic_slice_fusion_rewriter.cc

```diff
@@ -493,6 +496,10 @@ absl::StatusOr<bool> DynamicSliceFusionRewriter::Run(
       }
       TF_RETURN_IF_ERROR(
           parent->ReplaceInstruction(instr_to_be_replaced, fusion));
+      // This is required for collective operations which will not be removed.
+      if (hero->parent() && hero->user_count() == 0) {
+        TF_RETURN_IF_ERROR(hero->parent()->RemoveInstruction(hero));
```

Comment:
I think it is an error if after the rewrite `user_count` is not 0, and it should be safe to always remove instruction explicitly.